File: typescript-client-sdk/src/sdk/models/operations/postsubscriptions.ts
import { SpeakeasyBase, SpeakeasyMetadata } from "../../../internal/utils";
import * as shared from "../shared";
import { Expose, Transform, Type } from "class-transformer";

export enum PostSubscriptionsApplicationJSONExternalMarketplaceEnum {
    Google = "google",
    Aws = "aws",
    Azure = "azure"
}

export class PostSubscriptionsApplicationJSONPhaseOverrides extends SpeakeasyBase {
  @SpeakeasyMetadata()
  @Expose({ name: "discount" })
  discount?: Record<string, any>;

  @SpeakeasyMetadata()
  @Expose({ name: "minimum_amount" })
  minimumAmount?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "order" })
  order?: number;
}

export class PostSubscriptionsApplicationJSON extends SpeakeasyBase {
  @SpeakeasyMetadata()
  @Expose({ name: "align_billing_with_subscription_start_date" })
  alignBillingWithSubscriptionStartDate?: boolean;

  @SpeakeasyMetadata()
  @Expose({ name: "auto_collection" })
  autoCollection?: boolean;

  @SpeakeasyMetadata()
  @Expose({ name: "coupon_redemption_code" })
  couponRedemptionCode?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "customer_id" })
  customerId?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "default_invoice_memo" })
  defaultInvoiceMemo?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "external_customer_id" })
  externalCustomerId?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "external_marketplace" })
  externalMarketplace?: PostSubscriptionsApplicationJSONExternalMarketplaceEnum;

  @SpeakeasyMetadata()
  @Expose({ name: "external_marketplace_reporting_id" })
  externalMarketplaceReportingId?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "external_plan_id" })
  externalPlanId?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "minimum_amount" })
  minimumAmount?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "net_terms" })
  netTerms?: number;

  @SpeakeasyMetadata({ elemType: PostSubscriptionsApplicationJSONPhaseOverrides })
  @Expose({ name: "phase_overrides" })
  @Type(() => PostSubscriptionsApplicationJSONPhaseOverrides)
  phaseOverrides?: PostSubscriptionsApplicationJSONPhaseOverrides[];

  @SpeakeasyMetadata()
  @Expose({ name: "plan_id" })
  planId?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "price_overrides" })
  priceOverrides?: Record<string, any>[];

  @SpeakeasyMetadata()
  @Expose({ name: "start_date" })
  @Transform(({ value }) => new Date(value), { toClassOnly: true })
  startDate?: Date;
}

export class PostSubscriptionsRequests extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "request, media_type=application/xml" })
  applicationXML: Uint8Array;

  @SpeakeasyMetadata({ data: "request, media_type=application/json" })
  object?: PostSubscriptionsApplicationJSON;
}

export class PostSubscriptionsRequest extends SpeakeasyBase {
  @SpeakeasyMetadata()
  request?: PostSubscriptionsRequests;
}

export class PostSubscriptionsResponse extends SpeakeasyBase {
  @SpeakeasyMetadata()
  contentType: string;

  @SpeakeasyMetadata()
  statusCode: number;

  @SpeakeasyMetadata()
  subscription?: shared.Subscription;
}

File: python-client-sdk/src/orbapi/models/operations/post_subscriptions.py
from __future__ import annotations
import dataclasses
import dateutil.parser
from ..shared import subscription as shared_subscription
from dataclasses_json import Undefined, dataclass_json
from datetime import date
from enum import Enum
from marshmallow import fields
from orbapi import utils
from typing import Any, Optional

class PostSubscriptionsApplicationJSONExternalMarketplaceEnum(str, Enum):
    GOOGLE = "google"
    AWS = "aws"
    AZURE = "azure"


@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class PostSubscriptionsApplicationJSONPhaseOverrides:
    discount: Optional[dict[str, Any]] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('discount'), 'exclude': lambda f: f is None }})
    minimum_amount: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('minimum_amount'), 'exclude': lambda f: f is None }})
    order: Optional[float] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('order'), 'exclude': lambda f: f is None }})
    

@dataclass_json(undefined=Undefined.EXCLUDE)
@dataclasses.dataclass
class PostSubscriptionsApplicationJSON:
    align_billing_with_subscription_start_date: Optional[bool] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('align_billing_with_subscription_start_date'), 'exclude': lambda f: f is None }})
    auto_collection: Optional[bool] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('auto_collection'), 'exclude': lambda f: f is None }})
    coupon_redemption_code: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('coupon_redemption_code'), 'exclude': lambda f: f is None }})
    customer_id: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('customer_id'), 'exclude': lambda f: f is None }})
    default_invoice_memo: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('default_invoice_memo'), 'exclude': lambda f: f is None }})
    external_customer_id: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('external_customer_id'), 'exclude': lambda f: f is None }})
    external_marketplace: Optional[PostSubscriptionsApplicationJSONExternalMarketplaceEnum] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('external_marketplace'), 'exclude': lambda f: f is None }})
    external_marketplace_reporting_id: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('external_marketplace_reporting_id'), 'exclude': lambda f: f is None }})
    external_plan_id: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('external_plan_id'), 'exclude': lambda f: f is None }})
    minimum_amount: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('minimum_amount'), 'exclude': lambda f: f is None }})
    net_terms: Optional[int] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('net_terms'), 'exclude': lambda f: f is None }})
    phase_overrides: Optional[list[PostSubscriptionsApplicationJSONPhaseOverrides]] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('phase_overrides'), 'exclude': lambda f: f is None }})
    plan_id: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('plan_id'), 'exclude': lambda f: f is None }})
    price_overrides: Optional[list[dict[str, Any]]] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('price_overrides'), 'exclude': lambda f: f is None }})
    start_date: Optional[date] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('start_date'), 'encoder': utils.dateisoformat(True), 'decoder': utils.datefromisoformat, 'mm_field': fields.DateTime(format='iso'), 'exclude': lambda f: f is None }})
    

@dataclasses.dataclass
class PostSubscriptionsRequests:
    application_xml: bytes = dataclasses.field(metadata={'request': { 'media_type': 'application/xml' }})
    object: Optional[PostSubscriptionsApplicationJSON] = dataclasses.field(default=None, metadata={'request': { 'media_type': 'application/json' }})
    

@dataclasses.dataclass
class PostSubscriptionsRequest:
    request: Optional[PostSubscriptionsRequests] = dataclasses.field(default=None)
    

@dataclasses.dataclass
class PostSubscriptionsResponse:
    content_type: str = dataclasses.field()
    status_code: int = dataclasses.field()
    subscription: Optional[shared_subscription.Subscription] = dataclasses.field(default=None)
    